packages/tools/devtools/devtools-core/src/data-visualization/SharedTreeVisualizer.ts

```diff
@@ -148,64 +155,113 @@ export function toVisualTree(tree: VisualSharedTreeNode): VisualChildNode {
  */
-function getObjectAllowedTypes(schema: SimpleObjectNodeSchema): string {
-	const result: string[] = [];
+interface FieldSchemaProperties {
+	/**
+	 * Set of type names that are valid for this specific node position in the tree.
+	 * This is a subset of the types defined in treeDefinitions.
+	 */
+	allowedTypes?: ReadonlySet<string>;
 
-	for (const [fieldKey, treeFieldSimpleSchema] of Object.entries(schema.fields)) {
-		const fieldTypes = treeFieldSimpleSchema.allowedTypes;
-		result.push(`${fieldKey} : ${concatenateTypes(fieldTypes)}`);
-	}
-
-	return `{ ${result.join(", ")} }`;
+	/**
+	 * Whether this field is required in its parent object schema.
+	 * Only meaningful for direct children of object nodes.
+	 * Undefined for array/map elements since they are always required within their parent.
+	 */
+	isRequired?: boolean;
 }
 
 /**
- * Returns the schema & fields of the node.
+ * Processes and visualizes the fields of a verbose tree node.
+ *
+ * @param treeFields - The fields of the tree node to visualize. Can be either an array of VerboseTree (for array nodes) or a Record of field names to VerboseTree (for object/map nodes).
+ * @param treeDefinitions - Map containing all schema definitions for the entire tree structure. Each definition describes the shape and constraints of a particular node type.
+ * @param allowedTypes - Record mapping field names to their allowed type sets. Specifies which types are valid for each field in the tree.
+ * @param requirements - Optional record mapping field names to boolean values indicating whether each field is required (true) or optional (false). Only meaningful for object node fields.
+ *
+ * @returns A record mapping field names/indices to their visual tree representations.
  */
 async function visualizeVerboseNodeFields(
-	tree: VerboseTreeNode,
-	treeSchema: SimpleTreeSchema,
+	treeFields: readonly VerboseTree[] | Record<string, VerboseTree>,
+	treeDefinitions: ReadonlyMap<string, SimpleNodeSchema>,
+	allowedTypes: Record<string, ReadonlySet<string>>,
+	requirements: Record<string, boolean> | undefined,
 	visualizeChildData: VisualizeChildData,
 ): Promise<Record<string, VisualSharedTreeNode>> {
-	const treeFields = tree.fields;
-
 	const fields: Record<string | number, VisualSharedTreeNode> = {};
 
 	for (const [fieldKey, childField] of Object.entries(treeFields)) {
-		fields[fieldKey] = await visualizeSharedTreeNodeBySchema(
+		fields[fieldKey] = await visualizeSharedTreeBySchema(
 			childField,
-			treeSchema,
+			treeDefinitions,
+			{
+				allowedTypes: allowedTypes[fieldKey],
+				isRequired: requirements === undefined ? undefined : requirements[fieldKey],
+			},
 			visualizeChildData,
 		);
 	}
 
 	return fields;
 }
 
+/**
+ * Extracts and stores allowed types & kind from {@link SimpleFieldSchema}.
+ */
+function getFieldTooltipProperties(schema: SimpleObjectNodeSchema): {
+	allowedTypes: Record<string, ReadonlySet<string>>;
+	elementIsRequired: Record<string, boolean>;
```

Comment:
Nit: I think you could just call this `required` or `isRequired`. "Element" doesn't really narrow the description much IMO